semgrep-core/src/fixing/Autofix_metavar_replacement.ml

```diff
@@ -0,0 +1,125 @@
+(* Nat Mote
+ *
+ * Copyright (C) 2019-2022 r2c
+ *
+ * This library is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public License
+ * version 2.1 as published by the Free Software Foundation, with the
+ * special exception on linking described in file LICENSE.
+ *
+ * This library is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the file
+ * LICENSE for more details.
+ *)
+
+open AST_generic
+module MV = Metavariable
+
+(******************************************************************************)
+(* Module responsible for traversing a fix pattern AST and replacing
+ * metavariables within it with the AST nodes from the target file to which the
+ * metavariables are bound. *)
+(******************************************************************************)
+
+let make_metavar_tbl bindings =
+  let tbl = Hashtbl.create (List.length bindings) in
+  List.iter (fun (mvar, mvalue) -> Hashtbl.replace tbl mvar mvalue) bindings;
+  tbl
+
+let replace metavar_tbl pattern_ast =
+  (* Use a mapper to traverse the AST. For each metavar, look up what it is
+   * bound to. If the kind of node matches, replace the metavar in the pattern
+   * with the AST node to which it is bound. Note that we can't handle this by
+   * simply visiting the identifier nodes, because metavars are often bound to
+   * nodes other than identifiers. Instead, we have to inspect certain kinds of
+   * nodes for metavars and then replace the entire node. *)
+  let mapper =
+    Map_AST.(
+      mk_visitor
+        {
+          default_visitor with
+          (* TODO handle more kinds of nodes *)
+          kargs =
+            (fun (k, _) args ->
+              (* A metavariable can appear as a single argument, but can be
+               * bound to multiple arguments, so we have to handle this case by
+               * looking at an argument list as a whole *)
+              let map_arg arg =
+                match arg with
+                | Arg { e = N (Id ((id_str, _), _)); _ } -> (
+                    match Hashtbl.find_opt metavar_tbl id_str with
+                    | Some (MV.Args args) -> args
+                    | _ -> [ arg ])
+                | _ -> [ arg ]
+              in
+              let args = List.concat_map map_arg args in
+              k args);
+          kexpr =
+            (fun (k, _) orig ->
+              (* The mapper changes the ID of expressions. This stymies the
+               * print-avoidance implemented later in the autofix process,
+               * because the resulting AST nodes cannot be recognized as the
+               * same as the original ones.
+               *
+               * To work around this, we use the original expression, except for
```

Comment:
Another reason to do this: saving propagated values for later use in the message (via `value($X)`). So the workaround would still be necessary!

Reply:
Oh wait this isn't true since this is only used for synthesizing the fix text. I'll leave this comment as evidence of confusion